File: converter/java/src/main/java/com/mlt/decoder/IntegerDecoder.java
package com.mlt.decoder;

import com.mlt.converter.encodings.PhysicalLevelTechnique;
import com.mlt.converter.encodings.StreamMetadata;
import com.mlt.converter.geometry.ZOrderCurve;
import me.lemire.integercompression.IntWrapper;
import org.checkerframework.checker.units.qual.A;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class IntegerDecoder {

    private IntegerDecoder(){}

    public static List<Integer> decodeMortonStream(byte[] data, IntWrapper offset, StreamMetadata streamMetadata){
        int[] values;
        if(streamMetadata.physicalLevelTechnique() == PhysicalLevelTechnique.FAST_PFOR){
            //TODO: numValues is not right if rle or delta rle is used -> add separate flag in StreamMetadata
            values = DecodingUtils.decodeFastPfor128(data, streamMetadata.numValues(), streamMetadata.byteLength(), offset);
        }
        else if(streamMetadata.physicalLevelTechnique() == PhysicalLevelTechnique.VARINT){
            values = DecodingUtils.decodeVarint(data, offset, streamMetadata.numValues());
        }
        else{
            throw new IllegalArgumentException("Specified physical level technique not yet supported.");
        }

        return decodeMortonDelta(values, streamMetadata.numBits(), streamMetadata.coordinateShift());
    }

    private static List<Integer> decodeMortonDelta(int[] data, int numBits, int coordinateShift){
        var vertices = new ArrayList<Integer>(data.length * 2);
        var previousMortonCode = 0;
        for(var deltaCode : data){
            var mortonCode = previousMortonCode + deltaCode;
            var vertex = decodeMortonCode(mortonCode, numBits, coordinateShift);
            vertices.add(vertex[0]);
            vertices.add(vertex[1]);
            previousMortonCode = mortonCode;
        }

        return vertices;
    }

    private static int[] decodeMortonCode(int mortonCode, int numBits, int coordinateShift) {
        int x = decodeMorton(mortonCode, numBits) - coordinateShift;
        int y = decodeMorton(mortonCode >> 1, numBits) - coordinateShift;
        return new int[]{x, y};
    }

    private static int decodeMorton(int code, int numBits) {
        int coordinate = 0;
        for (int i = 0; i < numBits; i++) {
            coordinate |= (code & (1L << (2 * i))) >> i;
        }
        return coordinate;
    }

    public static List<Integer> decodeIntStream(byte[] data, IntWrapper offset, StreamMetadata streamMetadata, boolean isSigned){
        int[] values = null;
        if(streamMetadata.physicalLevelTechnique() == PhysicalLevelTechnique.FAST_PFOR){
            //TODO: numValues is not right if rle or delta rle is used -> add separate flag in StreamMetadata
            values = DecodingUtils.decodeFastPfor128(data, streamMetadata.numValues(), streamMetadata.byteLength(), offset);
        }
        else if(streamMetadata.physicalLevelTechnique() == PhysicalLevelTechnique.VARINT){
            values = DecodingUtils.decodeVarint(data, offset, streamMetadata.numValues());
        }
        else{
            throw new IllegalArgumentException("Specified physical level technique not yet supported.");
        }

        switch (streamMetadata.logicalLevelTechnique()){
            case RLE:{
                var decodedValues = decodeRLE(values, streamMetadata.numRunValues().get());
                return isSigned? decodeZigZag(decodedValues.stream().mapToInt(i -> i).toArray()) : decodedValues;
            }
            case DELTA:
                return decodeZigZagDelta(values);
            case DELTA_RLE:
                return decodeDeltaRLE(values, streamMetadata.numRunValues().get());
            case NONE:{
                var decodedValues = Arrays.stream(values).boxed().collect(Collectors.toList());
                return isSigned? decodeZigZag(decodedValues.stream().mapToInt(i -> i).toArray()) : decodedValues;
            }
            default:
                throw new IllegalArgumentException("The specified logical level technique is not supported for integers.");
        }
    }

    public static List<Long> decodeLongStream(byte[] data, IntWrapper offset, StreamMetadata streamMetadata, boolean isSigned){
        if(streamMetadata.physicalLevelTechnique() != PhysicalLevelTechnique.VARINT){
            throw new IllegalArgumentException("Specified physical level technique not yet supported.");
        }

        var values = DecodingUtils.decodeLongVarint(data, offset, streamMetadata.numValues());

        switch (streamMetadata.logicalLevelTechnique()){
            case RLE:{
                var decodedValues =  decodeRLE(values, streamMetadata.numRunValues().get());
                return isSigned? decodeZigZag(decodedValues.stream().mapToLong(i -> i).toArray()) : decodedValues;
            }
            case DELTA:
                return decodeZigZagDelta(values);
            case DELTA_RLE:{
                return decodeDeltaRLE(values, streamMetadata.numRunValues().get());
            }
            case NONE:{
                var decodedValues = Arrays.stream(values).boxed().collect(Collectors.toList());
                return isSigned? decodeZigZag(decodedValues.stream().mapToLong(i -> i).toArray()) : decodedValues;
            }
            default:
                throw new IllegalArgumentException("The specified logical level technique is not supported for integers.");
        }
    }

    //TODO: quick and dirty -> write fast vectorized solution
    private static List<Integer> decodeRLE(int[] data, int numRuns){
        //TODO: add numRleValues property to StreamMetadata to allocate values array
        var values = new ArrayList<Integer>();
        for(var i = 0; i < numRuns; i++){
            var run = data[i];
            var value = data[i + numRuns];
            for(var j = 0; j < run; j++){
                values.add(value);
            }
        }

        return values;
    }

    //TODO: quick and dirty -> write fast vectorized solution
    private static List<Long> decodeRLE(long[] data, int numRuns){
        var values = new ArrayList<Long>();
        for(var i = 0; i < numRuns; i++){
            var run = data[i];
            var value = data[i + numRuns];
            for(var j = 0; j < run; j++){
                values.add(value);
            }
        }

        return values;
    }

    private static List<Integer> decodeDeltaRLE(int[] data, int numRuns){
        var deltaValues = new ArrayList<Integer>();
        for(var i = 0; i < numRuns; i++){
            var run = data[i];
            /* Only values are zig-zag encoded */
            var delta = DecodingUtils.decodeZigZag(data[i + numRuns]);
            //values.add(delta + previousValue);
            for(var j = 0; j < run; j++){
                deltaValues.add(delta);
            }
        }

        //TODO: merge rle and delta encoding
        var values = new ArrayList<Integer>(deltaValues.size());
        var previousValue = 0;
        for(var delta : deltaValues){
            var value = delta + previousValue;
            values.add(value);
            previousValue = value;
        }

        return values;
    }

    private static List<Long> decodeDeltaRLE(long[] data, int numRuns){
        var deltaValues = new ArrayList<Long>();
        for(var i = 0; i < numRuns; i++){
            var run = data[i];
            /* Only values are zig-zag encoded */
            var delta = DecodingUtils.decodeZigZag(data[i + numRuns]);
            //values.add(delta + previousValue);
            for(var j = 0; j < run; j++){
                deltaValues.add(delta);
            }
        }

        //TODO: merge rle and delta encoding
        var values = new ArrayList<Long>(deltaValues.size());
        var previousValue = 0l;
        for(var delta : deltaValues){
            var value = delta + previousValue;
            values.add(value);
            previousValue = value;
        }

        return values;
    }

    private static List<Integer> decodeZigZagDelta(int[] data){
        var values = new ArrayList<Integer>(data.length);
        var previousValue = 0;
        for(var zigZagDelta : data){
            var delta = DecodingUtils.decodeZigZag(zigZagDelta);
            var value = previousValue + delta;
            values.add(value);
            previousValue = value;
        }

        return values;
    }

    private static List<Long> decodeZigZagDelta(long[] data){
        var values = new ArrayList<Long>(data.length);
        var previousValue = 0l;
        for(var zigZagDelta : data){
            var delta = DecodingUtils.decodeZigZag(zigZagDelta);
            var value = previousValue + delta;
            values.add(value);
            previousValue = value;
        }

        return values;
    }

    private static List<Long> decodeZigZag(long[] data){
        var values = new ArrayList<Long>(data.length);
        for(var zigZagDelta : data){
            var value = DecodingUtils.decodeZigZag(zigZagDelta);
            values.add(value);
        }
        return values;
    }

    private static List<Integer> decodeZigZag(int[] data){
        var values = new ArrayList<Integer>(data.length);
        for(var zigZagDelta : data){
            var value = DecodingUtils.decodeZigZag(zigZagDelta);
            values.add(value);
        }
        return values;
    }
}

File: js/src/decoder/IntegerDecoder.ts
import { DecodingUtils } from './DecodingUtils';
import { IntWrapper } from './IntWrapper';
import { StreamMetadata } from '../metadata/stream/StreamMetadata';
import { LogicalLevelTechnique } from '../metadata/stream/LogicalLevelTechnique';
import { PhysicalLevelTechnique } from '../metadata/stream/PhysicalLevelTechnique';
import { MortonEncodedStreamMetadata } from '../metadata/stream/MortonEncodedStreamMetadata';
import { RleEncodedStreamMetadata } from '../metadata/stream/RleEncodedStreamMetadata';

class IntegerDecoder {

    public static decodeMortonStream(data: Uint8Array, offset: IntWrapper, streamMetadata: MortonEncodedStreamMetadata): Int32Array {
        let values: Int32Array;
        if (streamMetadata.physicalLevelTechnique() === PhysicalLevelTechnique.FAST_PFOR) {
            throw new Error("Specified physical level technique not yet supported: " + streamMetadata.physicalLevelTechnique());
            // TODO
            //values = DecodingUtils.decodeFastPfor128(data, streamMetadata.numValues(), streamMetadata.byteLength(), offset);
        } else if (streamMetadata.physicalLevelTechnique() === PhysicalLevelTechnique.VARINT) {
            values = DecodingUtils.decodeVarint(data, offset, streamMetadata.numValues());
        } else {
            throw new Error("Specified physical level technique not yet supported: " + streamMetadata.physicalLevelTechnique());
        }

        return this.decodeMortonDelta(values, streamMetadata.numBits(), streamMetadata.coordinateShift());
    }

    private static decodeMortonDelta(data: Int32Array, numBits: number, coordinateShift: number): Int32Array {
        const vertices = new Int32Array(data.length * 2);
        let previousMortonCode = 0;
        let counter = 0;
        for (const deltaCode of data) {
            const code = previousMortonCode + deltaCode;
            let x = 0;
            let y = 0;
            const codeY = code >> 1;
            for (let i = 0; i < numBits; i++) {
                x |= (code & (1 << (2 * i))) >> i;
                y |= (codeY & (1 << (2 * i))) >> i;
            }
            vertices[counter++] = x - coordinateShift;
            vertices[counter++] = y - coordinateShift;
            previousMortonCode = code;
        }
        return vertices;
    }

    private static decodeMortonCodes(data: Int32Array, numBits: number, coordinateShift: number): Int32Array {
        const vertices = new Int32Array(data.length * 2);
        let counter = 0;
        for (const code of data) {
            let x = 0;
            let y = 0;
            const codeY = code >> 1;
            for (let i = 0; i < numBits; i++) {
                x |= (code & (1 << (2 * i))) >> i;
                y |= (codeY & (1 << (2 * i))) >> i;
            }
            vertices[counter++] = x - coordinateShift;
            vertices[counter++] = y - coordinateShift;
        }
        return vertices;
    }

    public static decodeIntStream(data: Uint8Array, offset: IntWrapper, streamMetadata: StreamMetadata, isSigned: boolean): Int32Array {
        let values: Int32Array;
        if (streamMetadata.physicalLevelTechnique() === PhysicalLevelTechnique.FAST_PFOR) {
            throw new Error("Specified physical level technique not yet supported: " + streamMetadata.physicalLevelTechnique());
            // TODO
            //values = DecodingUtils.decodeFastPfor128(data, streamMetadata.numValues(), streamMetadata.byteLength(), offset);
        } else if (streamMetadata.physicalLevelTechnique() === PhysicalLevelTechnique.VARINT) {
            values = DecodingUtils.decodeVarint(data, offset, streamMetadata.numValues());
        } else {
            throw new Error("Specified physical level technique not yet supported: " + streamMetadata.physicalLevelTechnique());
        }
        return this.decodeIntArray(values, streamMetadata, isSigned);
    }

    private static decodeIntArray(values: Int32Array, streamMetadata: StreamMetadata, isSigned: boolean): Int32Array {
        switch (streamMetadata.logicalLevelTechnique1()) {
            case LogicalLevelTechnique.DELTA: {
                if (streamMetadata.logicalLevelTechnique2() === LogicalLevelTechnique.RLE) {
                    const rleMetadata = streamMetadata as RleEncodedStreamMetadata;
                    // TODO: stop mutating these values?
                    values =
                        DecodingUtils.decodeUnsignedRLE(
                            values, rleMetadata.runs(), rleMetadata.numRleValues());
                    return this.decodeZigZagDelta(values);
                }
                return this.decodeZigZagDelta(values);
            }
            case LogicalLevelTechnique.RLE: {
                const rleMetadata = streamMetadata as RleEncodedStreamMetadata;
                const decodedValues = this.decodeRLE(values, rleMetadata.runs());
                return isSigned ? this.decodeZigZag(decodedValues) : decodedValues;
            }
            case LogicalLevelTechnique.NONE: {
                return isSigned ? this.decodeZigZag(values) : values;
            }
            case LogicalLevelTechnique.MORTON: {
                const mortonMetadata = streamMetadata as MortonEncodedStreamMetadata;
                return this.decodeMortonCodes(values, mortonMetadata.numBits(), mortonMetadata.coordinateShift());
            }
            case LogicalLevelTechnique.COMPONENTWISE_DELTA: {
                DecodingUtils.decodeComponentwiseDeltaVec2(values);
                return values;
            }
            default:
                throw new Error("The specified logical level technique is not supported for integers: " + streamMetadata.logicalLevelTechnique1());
        }
    }

    public static decodeLongStream(data: Uint8Array, offset: IntWrapper, streamMetadata: StreamMetadata, isSigned: boolean): BigInt64Array {
        if (streamMetadata.physicalLevelTechnique() !== PhysicalLevelTechnique.VARINT) {
            throw new Error("Specified physical level technique not yet supported: " + streamMetadata.physicalLevelTechnique());
        }

        const values = DecodingUtils.decodeLongVarint(data, offset, streamMetadata.numValues());
        return this.decodeLongArray(values, streamMetadata, isSigned);
    }

    private static decodeLongArray(values: BigInt64Array, streamMetadata: StreamMetadata, isSigned: boolean): BigInt64Array {
        switch (streamMetadata.logicalLevelTechnique1()) {
            case LogicalLevelTechnique.DELTA: {
                if (streamMetadata.logicalLevelTechnique2() === LogicalLevelTechnique.RLE) {
                    const rleMetadata = streamMetadata as RleEncodedStreamMetadata;
                    values =
                        DecodingUtils.decodeUnsignedRLELong(
                            values, rleMetadata.runs(), rleMetadata.numRleValues());
                    return this.decodeLongZigZagDelta(values);
                }
                return this.decodeLongZigZagDelta(values);
            }
            case LogicalLevelTechnique.RLE: {
                const rleMetadata = streamMetadata as RleEncodedStreamMetadata;
                const decodedValues = this.decodeLongRLE(values, rleMetadata.runs());
                return isSigned ? this.decodeZigZagLong(decodedValues) : decodedValues;
            }
            case LogicalLevelTechnique.NONE: {
                return isSigned ? this.decodeZigZagLong(values) : values;
            }
            default:
                throw new Error("The specified logical level technique is not supported for integers: " + streamMetadata.logicalLevelTechnique1());
        }
    }

    private static decodeRLE(data: Int32Array, numRuns: number): Int32Array {
        // Note: if this array is initialied like new Array<number>(numRleValues)
        // like the java implementation does, the array will potentially contain
        // extra uninitialized values
        // TODO: figure out how to pre-initialize the array with the correct size
        const values = new Array<number>();
        for (let i = 0; i < numRuns; i++) {
            const run = data[i];
            const value = data[i + numRuns];
            for (let j = 0; j < run; j++) {
                values.push(value);
            }
        }
        return new Int32Array(values);
    }

    private static decodeLongRLE(data: BigInt64Array, numRuns: number): BigInt64Array {
        // Note: if this array is initialied like new Array<number>(numRleValues)
        // like the java implementation does, the array will potentially contain
        // extra uninitialized values
        const values = new Array<bigint>();
        for (let i = 0; i < numRuns; i++) {
            const run = data[i];
            const value = data[i + numRuns];
            for (let j = 0; j < run; j++) {
                values.push(value);
            }
        }
        return new BigInt64Array(values);
    }

    private static decodeZigZagDelta(data: Int32Array): Int32Array {
        const values = new Int32Array(data.length);
        let previousValue = 0;
        let counter = 0;
        for (const zigZagDelta of data) {
            const value = previousValue + DecodingUtils.decodeZigZag(zigZagDelta);
            values[counter++] = value;
            previousValue = value;
        }
        return values;
    }

    private static decodeLongZigZagDelta(data: BigInt64Array): BigInt64Array {
        const values = new BigInt64Array(data.length);
        let previousValue = 0n;
        let counter = 0;
        for (const zigZagDelta of data) {
            const value = previousValue + DecodingUtils.decodeZigZagLong(zigZagDelta);
            values[counter++] = value;
            previousValue = value;
        }
        return values;
    }

    // TODO: check if mutating in place would be faster than returning a new array
    private static decodeZigZag(data: Int32Array): Int32Array {
        return data.map(zigZagDelta => DecodingUtils.decodeZigZag(zigZagDelta));
    }

    private static decodeZigZagLong(data: BigInt64Array): BigInt64Array {
        return data.map(zigZagDelta => DecodingUtils.decodeZigZagLong(zigZagDelta));
    }
}

export { IntegerDecoder };
